File: frontend/src/types/event.ts
export interface Event {
  id: number;
  title: string;
  description: string;
  startDateTime: string;
  endDateTime: string;
  user: {
    id: number;
    email: string;
  };
  // Recurrence fields
  recurrenceRule?: string;
  recurrenceEndDate?: string;
  recurrenceCount?: number;
  parentEventId?: number;
  originalStartDateTime?: string;
}

export interface CreateEventRequest {
  title: string;
  description: string;
  startDateTime: string;
  endDateTime: string;
  // Recurrence fields
  recurrenceRule?: string;
  recurrenceEndDate?: string;
  recurrenceCount?: number;
  // Tag IDs to associate with this event
  tagIds?: number[];
}

export interface EventFormData {
  title: string;
  description: string;
  startDateTime: string;
  endDateTime: string;
  // Recurrence fields
  repeatType: 'none' | 'daily' | 'weekly' | 'monthly';
  repeatEndType: 'never' | 'date' | 'count';
  repeatEndDate: string;
  repeatCount: number;
}

File: backend/src/main/java/com/example/calendar/model/Event.java
package com.example.calendar.model;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "events")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Event {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private LocalDateTime startDateTime;

    @Column(nullable = false)
    private LocalDateTime endDateTime;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    // Recurrence fields
    @Column(columnDefinition = "TEXT")
    private String recurrenceRule; // RFC 5545 RRULE string

    @Column
    private LocalDateTime recurrenceEndDate;

    @Column
    private Integer recurrenceCount;

    // For tracking individual occurrences of recurring events
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_event_id")
    private Event parentEvent;

    @Column
    private LocalDateTime originalStartDateTime; // Original start time for this occurrence

    // For handling exclusions in recurring events
    @Column(columnDefinition = "TEXT")
    private String excludedDates; // Comma-separated list of excluded dates in ISO format

    // Many-to-many relationship with tags
    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE}, fetch = FetchType.LAZY)
    @JoinTable(
        name = "event_tags",
        joinColumns = @JoinColumn(name = "event_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @Builder.Default
    private Set<Tag> tags = new HashSet<>();

    // Custom getter to return defensive copy and avoid EI_EXPOSE_REP
    public Set<Tag> getTags() {
        return tags == null ? new HashSet<>() : new HashSet<>(tags);
    }
}
